File: 230-kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.cpp
// Iterative Solution
// Time complexity = O(n) ; as we are traversing every element in the event that k = largest element of the BST

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

 #include <stack>      // could've used vector also

class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        stack<TreeNode *> st;
        TreeNode *curr = root;
        int n = 0;

        while(curr != NULL || st.size() != 0){
            while(curr != NULL){
                st.push(curr);
                curr = curr->left;
            }

            curr = st.top();
            st.pop();
            n += 1;
            
            if(n == k){
                return curr->val;
            }

            curr = curr->right;
        }
        return 0;
    }
};

File: 230-kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.py
# Iterative Solution
# Time complexity = O(n) ; as we are traversing every element in the event that k = largest element of the BST

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        stack = []
        curr = root
        n = 0          # will be used to compare with k 

        while curr or stack:
            while curr:
                stack.append(curr)
                curr = curr.left
            
            curr = stack.pop()
            n += 1

            if n == k:
                return curr.val
            
            curr = curr.right
        
        